### PR TITLE
[SPARK-10974][Streaming]Add progress bar for output operation column and use red dots for failed batches

### PR DESCRIPTION
**Pull request by zsxwing**
Screenshot:

Also fixed the description and duration for output operations that don't have spark jobs.

### CODE REVIEW

**Comment by zsxwing** [comment]
Screenshot for red dots:

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala [hunk 1]
```diff
@@ -36,13 +37,10 @@ case class BatchInfo(
     streamIdToInputInfo: Map[Int, StreamInputInfo],
     submissionTime: Long,
     processingStartTime: Option[Long],
-    processingEndTime: Option[Long]
+    processingEndTime: Option[Long],
+    numOutputOp: Int
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala]: I think instead of the `numOutputOp`, we need to give references to the OutputOperationInfo. Kind of link SparkListenerJobStart event has StageInfos in it.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala [hunk 2]
```diff
@@ -29,7 +31,7 @@ class Job(val time: Time, func: () => _) {
   private var _outputOpId: Int = _
   private var isSet = false
   private var _result: Try[_] = null
-  private var _callSite: String = "Unknown"
+  private var _callSite: CallSite = null
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: why was this needed?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 3]
```diff
@@ -163,15 +163,35 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
       listenerBus.post(StreamingListenerBatchStarted(jobSet.toBatchInfo))
     }
     listenerBus.post(StreamingListenerOutputOperationStarted(
-      OutputOperationInfo(job.time, job.outputOpId, job.callSite, Some(startTime), None)))
+      OutputOperationInfo(
+        job.time,
+        job.outputOpId,
+        job.callSite.shortForm,
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: Why os the shortForm mapped to name and longForm mapped to description in OutputOperationInfo?

**Comment by tdas** [comment]
overall looks pretty good! I have one main comment that its kind of weird just expose the `numOuptutOp`. If we have to expose something, we should expose the whole information about output oeprations, especially they are designed for being exposed.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala (hunk 3 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: This is similar to `org.apache.spark.scheduler.Stage`:
```
abstract class Stage {
 ...
  val name = callSite.shortForm
  val details = callSite.longForm
...
}
```

Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: And `StageInfo` gets its `name` and `details` just from `Stage.name` and `Stage.details`.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala (hunk 2 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: See my comment bellow

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala (hunk 1 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala]: Good point. I will update it.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala [hunk 4]
```diff
@@ -71,11 +75,29 @@ class Job(val time: Time, func: () => _) {
...
+    _startTime = Some(startTime)
+  }
+
+  def setEndTime(endTime: Long): Unit = {
+    _endTime = Some(endTime)
+  }
+
+  def toOutputOperationInfo: OutputOperationInfo = {
+    val failureReason = if (_result != null && _result.isFailure) {
+        Some(Utils.exceptionString(_result.asInstanceOf[Failure[_]].exception))
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: Nit: Isnt this incorrectly indented? Shouldnt this be indented 2 spaces instead of 4?

**Comment by tdas** [comment]
LGTM, just one minor comment.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala (hunk 4 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: Fixed

**Comment by tdas** [comment]
Cool! I am merging this to master! Thanks @zsxwing
